File: ios/TrueSheetView.m
//
//  Created by Jovanni Lo (@lodev09)
//  Copyright (c) 2024-present. All rights reserved.
//
//  This source code is licensed under the MIT license found in the
//  LICENSE file in the root directory of this source tree.
//

#import "TrueSheetView.h"
#import "TrueSheetViewController.h"
#import "TrueSheetEvent.h"
#import <React/RCTUIManager.h>
#import <React/RCTTouchHandler.h>
#import <React/RCTSurfaceTouchHandler.h>
#import <React/RCTEventDispatcher.h>
#import <React/RCTConvert.h>
#import <React/UIView+React.h>

@interface TrueSheetView () <TrueSheetViewControllerDelegate>

@property (nonatomic, strong) RCTBridge *bridge;
@property (nonatomic, strong) id<RCTEventDispatcherProtocol> eventDispatcher;
@property (nonatomic, strong) TrueSheetViewController *viewController;
@property (nonatomic, strong) RCTTouchHandler *touchHandler;
@property (nonatomic, strong) RCTSurfaceTouchHandler *surfaceTouchHandler;

@property (nonatomic, assign) BOOL isPresented;
@property (nonatomic, strong, nullable) NSNumber *activeIndex;

@property (nonatomic, strong, nullable) UIView *containerView;
@property (nonatomic, strong, nullable) UIView *contentView;
@property (nonatomic, strong, nullable) UIView *footerView;
@property (nonatomic, strong, nullable) UIView *scrollView;

@property (nonatomic, strong, nullable) NSLayoutConstraint *footerBottomConstraint;
@property (nonatomic, strong, nullable) NSLayoutConstraint *footerHeightConstraint;

@end

@implementation TrueSheetView

- (instancetype)initWithBridge:(RCTBridge *)bridge {
    if (self = [super initWithFrame:CGRectZero]) {
        _bridge = bridge;
        _eventDispatcher = bridge.eventDispatcher;
        _viewController = [[TrueSheetViewController alloc] init];
        _viewController.delegate = self;
        _touchHandler = [[RCTTouchHandler alloc] initWithBridge:bridge];
        _surfaceTouchHandler = [[RCTSurfaceTouchHandler alloc] init];
        _isPresented = NO;
        _initialIndex = @(-1);
        _initialIndexAnimated = YES;
        _grabber = YES;
        _dismissible = YES;
        _dimmed = YES;
    }
    return self;
}

- (void)dealloc {
    [self invalidate];
}

- (void)invalidate {
    if (_isPresented) {
        [_viewController dismissViewControllerAnimated:YES completion:nil];
    }
}

#pragma mark - React Subview Management

- (void)insertReactSubview:(UIView *)subview atIndex:(NSInteger)atIndex {
    [super insertReactSubview:subview atIndex:atIndex];
    
    if (self.containerView != nil) {
        NSLog(@"TrueSheet: Sheet can only have one content view.");
        return;
    }
    
    self.containerView = subview;
    [self.viewController.view addSubview:subview];
    [self.touchHandler attachToView:subview];
    [self.surfaceTouchHandler attachToView:subview];
}

- (void)removeReactSubview:(UIView *)subview {
    if (subview != self.containerView) {
        NSLog(@"TrueSheet: Cannot remove view other than sheet view");
        return;
    }
    
    [super removeReactSubview:subview];
    
    [self.touchHandler detachFromView:subview];
    [self.surfaceTouchHandler detachFromView:subview];
    
    [self unpinView:self.containerView];
    [self unpinView:self.footerView];
    [self unpinView:self.contentView];
    [self unpinView:self.scrollView];
    
    self.containerView = nil;
    self.contentView = nil;
    self.footerView = nil;
    self.scrollView = nil;
}

- (void)didUpdateReactSubviews {
    // Do nothing, as subviews are managed by insertReactSubview
}

- (void)layoutSubviews {
    [super layoutSubviews];
    
    if (self.containerView != nil && self.contentView == nil) {
        if (self.containerView.subviews.count >= 1) {
            self.contentView = self.containerView.subviews[0];
        }
        if (self.containerView.subviews.count >= 2) {
            self.footerView = self.containerView.subviews[1];
        }
        
        [self pinView:self.containerView toView:self.viewController.view edges:UIRectEdgeAll];
        
        if (self.contentView) {
            CGFloat contentHeight = self.contentView.bounds.size.height;
            [self setContentHeight:@(contentHeight)];
        }
        
        if (self.footerView) {
            [self setupFooterConstraints];
            CGFloat footerHeight = self.footerView.bounds.size.height;
            [self setFooterHeight:@(footerHeight)];
        }
        
        NSInteger initialIndex = [self.initialIndex integerValue];
        if (initialIndex >= 0) {
            [self presentAtIndex:initialIndex animated:self.initialIndexAnimated resolve:nil reject:nil];
        }
        
        [self dispatchEventWithName:@"onMount" block:self.onMount data:nil];
    }
}

#pragma mark - Layout Helpers

- (void)pinView:(UIView *)view toView:(UIView *)parentView edges:(UIRectEdge)edges {
    view.translatesAutoresizingMaskIntoConstraints = NO;
    
    if (edges & UIRectEdgeTop) {
        [view.topAnchor constraintEqualToAnchor:parentView.topAnchor].active = YES;
    }
    if (edges & UIRectEdgeBottom) {
        [view.bottomAnchor constraintEqualToAnchor:parentView.bottomAnchor].active = YES;
    }
    if (edges & UIRectEdgeLeft) {
        [view.leadingAnchor constraintEqualToAnchor:parentView.leadingAnchor].active = YES;
    }
    if (edges & UIRectEdgeRight) {
        [view.trailingAnchor constraintEqualToAnchor:parentView.trailingAnchor].active = YES;
    }
}

- (void)setupFooterConstraints {
    if (!self.footerView) return;
    
    self.footerView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.footerView.leadingAnchor constraintEqualToAnchor:self.viewController.view.leadingAnchor].active = YES;
    [self.footerView.trailingAnchor constraintEqualToAnchor:self.viewController.view.trailingAnchor].active = YES;
    
    self.footerBottomConstraint = [self.footerView.bottomAnchor constraintEqualToAnchor:self.viewController.view.bottomAnchor];
    self.footerBottomConstraint.active = YES;
}

- (void)unpinView:(UIView *)view {
    if (!view) return;
    view.translatesAutoresizingMaskIntoConstraints = YES;
    [view removeConstraints:view.constraints];
}

#pragma mark - TrueSheetViewControllerDelegate

- (void)viewControllerKeyboardWillHide {
    self.footerBottomConstraint.constant = 0;
    [UIView animateWithDuration:0.3 animations:^{
        [self.viewController.view layoutIfNeeded];
    }];
}

- (void)viewControllerKeyboardWillShow:(CGFloat)keyboardHeight {
    self.footerBottomConstraint.constant = -keyboardHeight;
    [UIView animateWithDuration:0.3 animations:^{
        [self.viewController.view layoutIfNeeded];
    }];
}

- (void)viewControllerDidChangeWidth:(CGFloat)width {
    [self dispatchEventWithName:@"onContainerSizeChange" 
                          block:self.onContainerSizeChange 
                           data:@{@"width": @(width)}];
}

- (void)viewControllerDidDrag:(UIGestureRecognizerState)state height:(CGFloat)height {
    NSInteger index = self.activeIndex ? [self.activeIndex integerValue] : 0;
    NSDictionary *sizeInfo = @{@"index": @(index), @"value": @(height)};
    
    switch (state) {
        case UIGestureRecognizerStateBegan:
            [self dispatchEventWithName:@"onDragBegin" block:self.onDragBegin data:sizeInfo];
            break;
        case UIGestureRecognizerStateChanged:
            [self dispatchEventWithName:@"onDragChange" block:self.onDragChange data:sizeInfo];
            break;
        case UIGestureRecognizerStateEnded:
        case UIGestureRecognizerStateCancelled:
            [self dispatchEventWithName:@"onDragEnd" block:self.onDragEnd data:sizeInfo];
            break;
        default:
            NSLog(@"TrueSheet: Drag state is not supported");
            break;
    }
}

- (void)viewControllerWillAppear {
    if (self.contentView && self.scrollView && self.containerView) {
        [self pinView:self.contentView toView:self.containerView edges:UIRectEdgeAll];
        [self pinView:self.scrollView toView:self.contentView edges:UIRectEdgeAll];
    }
}

- (void)viewControllerDidDismiss {
    self.isPresented = NO;
    self.activeIndex = nil;
    [self dispatchEventWithName:@"onDismiss" block:self.onDismiss data:nil];
}

- (void)viewControllerDidChangeSize:(NSInteger)index value:(CGFloat)value {
    if (!self.activeIndex || [self.activeIndex integerValue] != index) {
        self.activeIndex = @(index);
        NSDictionary *sizeInfo = @{@"index": @(index), @"value": @(value)};
        [self dispatchEventWithName:@"onSizeChange" block:self.onSizeChange data:sizeInfo];
    }
}

#pragma mark - Property Setters

- (void)setDismissible:(BOOL)dismissible {
    _dismissible = dismissible;
    self.viewController.modalInPresentation = !dismissible;
}

- (void)setMaxHeight:(NSNumber *)maxHeight {
    if ([_maxHeight isEqualToNumber:maxHeight]) return;
    _maxHeight = maxHeight;
    self.viewController.maxHeight = maxHeight;
    
    if (@available(iOS 15.0, *)) {
        if (self.isPresented) {
            [self.viewController setupSizes];
        }
    }
}

- (void)setContentHeight:(NSNumber *)contentHeight {
    if ([_contentHeight isEqualToNumber:contentHeight]) return;
    _contentHeight = contentHeight;
    self.viewController.contentHeight = contentHeight;
    
    if (@available(iOS 15.0, *)) {
        if (self.isPresented) {
            [self.viewController setupSizes];
        }
    }
}

- (void)setFooterHeight:(NSNumber *)footerHeight {
    if (!self.footerView || [_footerHeight isEqualToNumber:footerHeight]) return;
    _footerHeight = footerHeight;
    self.viewController.footerHeight = footerHeight;
    
    if (self.footerView.subviews.firstObject != nil) {
        [self.containerView bringSubviewToFront:self.footerView];
        if (!self.footerHeightConstraint) {
            self.footerHeightConstraint = [self.footerView.heightAnchor constraintEqualToConstant:[footerHeight floatValue]];
            self.footerHeightConstraint.active = YES;
        } else {
            self.footerHeightConstraint.constant = [footerHeight floatValue];
        }
    } else {
        [self.containerView sendSubviewToBack:self.footerView];
        if (self.footerHeightConstraint) {
            self.footerHeightConstraint.constant = 0;
        }
    }
    
    if (@available(iOS 15.0, *)) {
        if (self.isPresented) {
            [self.viewController setupSizes];
        }
    }
}

- (void)setSizes:(NSArray *)sizes {
    _sizes = sizes;
    self.viewController.sizes = sizes;
    
    if (@available(iOS 15.0, *)) {
        if (self.isPresented) {
            [self.viewController setupSizes];
        }
    }
}

- (void)setBackground:(NSNumber *)background {
    _background = background;
    self.viewController.backgroundColor = [RCTConvert UIColor:background];
    [self.viewController setupBackground];
}

- (void)setBlurTint:(NSString *)blurTint {
    _blurTint = blurTint;
    if (blurTint) {
        self.viewController.blurEffect = [self blurEffectFromString:blurTint];
    } else {
        self.viewController.blurEffect = nil;
    }
    [self.viewController setupBackground];
}

- (void)setCornerRadius:(NSNumber *)cornerRadius {
    _cornerRadius = cornerRadius;
    self.viewController.cornerRadius = cornerRadius;
    
    if (@available(iOS 15.0, *)) {
        if (self.isPresented && self.viewController.sheetPresentationController) {
            UISheetPresentationController *sheet = self.viewController.sheetPresentationController;
            [sheet animateChanges:^{
                sheet.preferredCornerRadius = cornerRadius ? [cornerRadius floatValue] : 0;
            }];
        }
    }
}

- (void)setGrabber:(BOOL)grabber {
    _grabber = grabber;
    self.viewController.grabber = grabber;
    
    if (@available(iOS 15.0, *)) {
        if (self.isPresented && self.viewController.sheetPresentationController) {
            UISheetPresentationController *sheet = self.viewController.sheetPresentationController;
            [sheet animateChanges:^{
                sheet.prefersGrabberVisible = grabber;
            }];
        }
    }
}

- (void)setDimmed:(BOOL)dimmed {
    if (_dimmed == dimmed) return;
    _dimmed = dimmed;
    self.viewController.dimmed = dimmed;
    
    if (@available(iOS 15.0, *)) {
        if (self.isPresented) {
            [self.viewController setupDimmedBackground];
        }
    }
}

- (void)setDimmedIndex:(NSNumber *)dimmedIndex {
    if ([_dimmedIndex isEqualToNumber:dimmedIndex]) return;
    _dimmedIndex = dimmedIndex;
    self.viewController.dimmedIndex = dimmedIndex;
    
    if (@available(iOS 15.0, *)) {
        if (self.isPresented) {
            [self.viewController setupDimmedBackground];
        }
    }
}

- (void)setScrollableHandle:(NSNumber *)scrollableHandle {
    _scrollableHandle = scrollableHandle;
    self.scrollView = [self.bridge.uiManager viewForReactTag:scrollableHandle];
}

#pragma mark - Helper Methods

- (UIBlurEffect *)blurEffectFromString:(NSString *)tint {
    UIBlurEffectStyle style = UIBlurEffectStyleRegular;
    
    if ([tint isEqualToString:@"default"]) {
        style = UIBlurEffectStyleRegular;
    } else if ([tint isEqualToString:@"extraLight"]) {
        style = UIBlurEffectStyleExtraLight;
    } else if ([tint isEqualToString:@"light"]) {
        style = UIBlurEffectStyleLight;
    } else if ([tint isEqualToString:@"regular"]) {
        style = UIBlurEffectStyleRegular;
    } else if ([tint isEqualToString:@"dark"]) {
        style = UIBlurEffectStyleDark;
    } else if ([tint isEqualToString:@"prominent"]) {
        style = UIBlurEffectStyleProminent;
    } else if ([tint isEqualToString:@"systemUltraThinMaterial"]) {
        style = UIBlurEffectStyleSystemUltraThinMaterial;
    } else if ([tint isEqualToString:@"systemThinMaterial"]) {
        style = UIBlurEffectStyleSystemThinMaterial;
    } else if ([tint isEqualToString:@"systemMaterial"]) {
        style = UIBlurEffectStyleSystemMaterial;
    } else if ([tint isEqualToString:@"systemThickMaterial"]) {
        style = UIBlurEffectStyleSystemThickMaterial;
    } else if ([tint isEqualToString:@"systemChromeMaterial"]) {
        style = UIBlurEffectStyleSystemChromeMaterial;
    } else if ([tint isEqualToString:@"systemUltraThinMaterialLight"]) {
        style = UIBlurEffectStyleSystemUltraThinMaterialLight;
    } else if ([tint isEqualToString:@"systemThickMaterialLight"]) {
        style = UIBlurEffectStyleSystemThickMaterialLight;
    } else if ([tint isEqualToString:@"systemThinMaterialLight"]) {
        style = UIBlurEffectStyleSystemThinMaterialLight;
    } else if ([tint isEqualToString:@"systemMaterialLight"]) {
        style = UIBlurEffectStyleSystemMaterialLight;
    } else if ([tint isEqualToString:@"systemChromeMaterialLight"]) {
        style = UIBlurEffectStyleSystemChromeMaterialLight;
    } else if ([tint isEqualToString:@"systemUltraThinMaterialDark"]) {
        style = UIBlurEffectStyleSystemUltraThinMaterialDark;
    } else if ([tint isEqualToString:@"systemThinMaterialDark"]) {
        style = UIBlurEffectStyleSystemThinMaterialDark;
    } else if ([tint isEqualToString:@"systemMaterialDark"]) {
        style = UIBlurEffectStyleSystemMaterialDark;
    } else if ([tint isEqualToString:@"systemThickMaterialDark"]) {
        style = UIBlurEffectStyleSystemThickMaterialDark;
    } else if ([tint isEqualToString:@"systemChromeMaterialDark"]) {
        style = UIBlurEffectStyleSystemChromeMaterialDark;
    }
    
    return [UIBlurEffect effectWithStyle:style];
}

- (void)dispatchEventWithName:(NSString *)name block:(RCTDirectEventBlock)block data:(NSDictionary *)data {
#ifdef RCT_NEW_ARCH_ENABLED
    if (block) {
        block(data);
    }
#else
    TrueSheetEvent *event = [[TrueSheetEvent alloc] initWithViewTag:self.reactTag name:name data:data];
    [self.eventDispatcher sendEvent:event];
#endif
}

#pragma mark - Public Methods

- (void)presentAtIndex:(NSInteger)index resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject {
    [self presentAtIndex:index animated:YES resolve:resolve reject:reject];
}

- (void)presentAtIndex:(NSInteger)index animated:(BOOL)animated resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject {
    UIViewController *rvc = self.reactViewController;
    
    if (!rvc) {
        if (reject) {
            reject(@"Error", @"No react view controller present.", nil);
        }
        return;
    }
    
    if (index >= self.viewController.sizes.count) {
        if (reject) {
            reject(@"Error", [NSString stringWithFormat:@"Size at %ld is not configured.", (long)index], nil);
        }
        return;
    }
    
    if (self.isPresented) {
        if (@available(iOS 15.0, *)) {
            UISheetPresentationController *sheet = self.viewController.sheetPresentationController;
            if (sheet) {
                [sheet animateChanges:^{
                    sheet.selectedDetentIdentifier = [self.viewController detentIdentifierForIndex:index];
                    
                    NSDictionary *sizeInfo = [self.viewController currentSizeInfo];
                    if (sizeInfo) {
                        NSInteger currentIndex = [sizeInfo[@"index"] integerValue];
                        CGFloat value = [sizeInfo[@"value"] floatValue];
                        [self viewControllerDidChangeSize:currentIndex value:value];
                    }
                    
                    if (resolve) {
                        resolve(nil);
                    }
                }];
            }
        } else {
            NSDictionary *sizeInfo = [self.viewController currentSizeInfo];
            if (sizeInfo) {
                NSInteger currentIndex = [sizeInfo[@"index"] integerValue];
                CGFloat value = [sizeInfo[@"value"] floatValue];
                [self viewControllerDidChangeSize:currentIndex value:value];
            }
            if (resolve) {
                resolve(nil);
            }
        }
    } else {
        [self.viewController prepareForPresentationAtIndex:index completion:^{
            self.activeIndex = @(index);
            self.isPresented = YES;
            
            [rvc presentViewController:self.viewController animated:animated completion:^{
                if (@available(iOS 15.0, *)) {
                    [self.viewController observeDrag];
                }
                
                NSDictionary *sizeInfo = [self.viewController currentSizeInfo];
                [self dispatchEventWithName:@"onPresent" block:self.onPresent data:sizeInfo];
                
                if (resolve) {
                    resolve(nil);
                }
            }];
        }];
    }
}

- (void)dismissWithResolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject {
    if (!self.isPresented) {
        if (resolve) {
            resolve(nil);
        }
        return;
    }
    
    [self.viewController dismissViewControllerAnimated:YES completion:^{
        if (resolve) {
            resolve(nil);
        }
    }];
}

@end

File: ios/TrueSheetView.swift
//
//  Created by Jovanni Lo (@lodev09)
//  Copyright (c) 2024-present. All rights reserved.
//
//  This source code is licensed under the MIT license found in the
//  LICENSE file in the root directory of this source tree.
//

@objc(TrueSheetView)
class TrueSheetView: UIView, RCTInvalidating, TrueSheetViewControllerDelegate {
  // MARK: - Static properties

  // Keep track of active sheet opacity values (stack)
  private static var sheetOpacityStack: [CGFloat] = []

  // MARK: - React properties

  // MARK: - Events

  @objc var onMount: RCTDirectEventBlock?
  @objc var onDismiss: RCTDirectEventBlock?
  @objc var onPresent: RCTDirectEventBlock?
  @objc var onSizeChange: RCTDirectEventBlock?
  @objc var onContainerSizeChange: RCTDirectEventBlock?
  @objc var onDragBegin: RCTDirectEventBlock?
  @objc var onDragChange: RCTDirectEventBlock?
  @objc var onDragEnd: RCTDirectEventBlock?

  // MARK: - React Properties

  @objc var initialIndex: NSNumber = -1
  @objc var initialIndexAnimated = true
  @objc var dimmedAlpha: CGFloat = 0.75

  // MARK: - Private properties

  private var isPresented = false
  private var activeIndex: Int?
  private var bridge: RCTBridge?
  private var eventDispatcher: (any RCTEventDispatcherProtocol)?
  private var viewController: TrueSheetViewController

  private var touchHandler: RCTTouchHandler
  // New Arch
  private var surfaceTouchHandler: RCTSurfaceTouchHandler

  // MARK: - Content properties

  private var containerView: UIView?
  private var contentView: UIView?
  private var footerView: UIView?
  private var scrollView: UIView?

  // Bottom: Reference the bottom constraint to adjust during keyboard event
  // Height: Reference height constraint during content updates
  private var footerConstraints: Constraints?

  private var uiManager: RCTUIManager? {
    guard let uiManager = bridge?.uiManager else { return nil }
    return uiManager
  }

  // MARK: - Setup

  init(with bridge: RCTBridge) {
    self.bridge = bridge
    eventDispatcher = bridge.eventDispatcher()

    viewController = TrueSheetViewController()
    touchHandler = RCTTouchHandler(bridge: bridge)
    surfaceTouchHandler = RCTSurfaceTouchHandler()

    super.init(frame: .zero)

    viewController.delegate = self
  }

  @available(*, unavailable)
  required init?(coder _: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }

  override func insertReactSubview(_ subview: UIView!, at index: Int) {
    super.insertReactSubview(subview, at: index)

    guard containerView == nil else {
      Logger.error("Sheet can only have one content view.")
      return
    }

    containerView = subview

    viewController.view.addSubview(subview)
    touchHandler.attach(to: subview)
    surfaceTouchHandler.attach(to: subview)
  }

  override func removeReactSubview(_ subview: UIView!) {
    guard subview == containerView else {
      Logger.error("Cannot remove view other than sheet view")
      return
    }

    super.removeReactSubview(subview)

    // Touch handler for Old Arch
    touchHandler.detach(from: subview)

    // Touch handler that works in New Arch
    surfaceTouchHandler.detach(from: subview)

    // Remove all constraints
    // Fixes New Arch weird layout issue :/
    containerView?.unpin()
    footerView?.unpin()
    contentView?.unpin()
    scrollView?.unpin()

    containerView = nil
    contentView = nil
    footerView = nil
    scrollView = nil
  }

  override func didUpdateReactSubviews() {
    // Do nothing, as subviews are managed by `insertReactSubview`
  }

  override func layoutSubviews() {
    super.layoutSubviews()

    if let containerView, contentView == nil {
      contentView = containerView.subviews[0]
      footerView = containerView.subviews[1]

      containerView.pinTo(view: viewController.view, constraints: nil)

      if let contentView {
        // Set initial content height
        let contentHeight = contentView.bounds.height
        setContentHeight(NSNumber(value: contentHeight))
      }

      // Set footer constraints
      if let footerView {
        footerView.pinTo(view: viewController.view, from: [.left, .right, .bottom], with: 0) { constraints in
          self.footerConstraints = constraints
        }

        // Set initial footer height
        let footerHeight = footerView.bounds.height
        setFooterHeight(NSNumber(value: footerHeight))
      }

      // Present sheet at initial index
      let initialIndex = self.initialIndex.intValue
      if initialIndex >= 0 {
        present(at: initialIndex, promise: nil, animated: initialIndexAnimated)
      }

      dispatchEvent(name: "onMount", block: onMount, data: nil)
    }
  }

  // MARK: - ViewController delegate

  func viewControllerKeyboardWillHide() {
    footerConstraints?.bottom?.constant = 0

    UIView.animate(withDuration: 0.3) {
      self.viewController.view.layoutIfNeeded()
    }
  }

  func viewControllerKeyboardWillShow(_ keyboardHeight: CGFloat) {
    footerConstraints?.bottom?.constant = -keyboardHeight

    UIView.animate(withDuration: 0.3) {
      self.viewController.view.layoutIfNeeded()
    }
  }

  func viewControllerDidChangeWidth(_ width: CGFloat) {
    // We only pass width to JS since height is handled by the constraints
    dispatchEvent(name: "onContainerSizeChange", block: onContainerSizeChange, data: ["width": width])
  }

  func viewControllerDidDrag(_ state: UIGestureRecognizer.State, _ height: CGFloat) {
    let sizeInfo = SizeInfo(index: activeIndex ?? 0, value: height)

    switch state {
    case .began:
      dispatchEvent(name: "onDragBegin", block: onDragBegin, data: sizeInfoData(from: sizeInfo))
    case .changed:
      dispatchEvent(name: "onDragChange", block: onDragChange, data: sizeInfoData(from: sizeInfo))
    case .ended, .cancelled:
      dispatchEvent(name: "onDragEnd", block: onDragEnd, data: sizeInfoData(from: sizeInfo))
    default:
      Logger.info("Drag state is not supported")
    }
  }

  func viewControllerWillAppear() {
    // Only apply dimming if the sheet has dimmed property set to true
    if viewController.dimmed {
      let opacity = 1 - dimmedAlpha
      // Add this sheet's opacity to the stack and dim root view controller
      TrueSheetView.sheetOpacityStack.append(opacity)
      UIView.animate(withDuration: 0.3) {
        if let rootViewController = UIApplication.shared.windows.first?.rootViewController {
          rootViewController.view.alpha = opacity
        }
      }
    }

    guard let contentView, let scrollView, let containerView else {
      return
    }

    // Add constraints to fix weirdness and support ScrollView
    contentView.pinTo(view: containerView, constraints: nil)
    scrollView.pinTo(view: contentView, constraints: nil)
  }

  func viewControllerWillDisappear() {
    if viewController.isBeingDismissed && viewController.dimmed {
      if !TrueSheetView.sheetOpacityStack.isEmpty {
        TrueSheetView.sheetOpacityStack.removeLast()
      }

      UIView.animate(withDuration: 0.3) {
        if let rootViewController = UIApplication.shared.windows.first?.rootViewController {
          // If there are still active sheets, apply the alpha of the topmost one
          if let topmostOpacity = TrueSheetView.sheetOpacityStack.last {
            rootViewController.view.alpha = topmostOpacity
          } else {
            // If no sheets are left, restore alpha to 1
            rootViewController.view.alpha = 1
          }
        }
      }
    }
  }

  func viewControllerDidDismiss() {
    isPresented = false
    activeIndex = nil
    dispatchEvent(name: "onDismiss", block: onDismiss, data: nil)
  }

  func viewControllerDidChangeSize(_ sizeInfo: SizeInfo?) {
    guard let sizeInfo else { return }

    if sizeInfo.index != activeIndex {
      activeIndex = sizeInfo.index
      dispatchEvent(name: "onSizeChange", block: onSizeChange, data: sizeInfoData(from: sizeInfo))
    }
  }

  func invalidate() {
    viewController.dismiss(animated: true)
  }

  // MARK: - Prop setters

  @objc
  func setDismissible(_ dismissible: Bool) {
    viewController.isModalInPresentation = !dismissible
  }

  @objc
  func setMaxHeight(_ height: NSNumber) {
    let maxHeight = CGFloat(height.floatValue)
    guard viewController.maxHeight != maxHeight else {
      return
    }

    viewController.maxHeight = maxHeight

    if #available(iOS 15.0, *) {
      withPresentedSheet { _ in
        viewController.setupSizes()
      }
    }
  }

  @objc
  func setContentHeight(_ height: NSNumber) {
    let contentHeight = CGFloat(height.floatValue)
    guard viewController.contentHeight != contentHeight else {
      return
    }

    viewController.contentHeight = contentHeight

    if #available(iOS 15.0, *) {
      withPresentedSheet { _ in
        viewController.setupSizes()
      }
    }
  }

  @objc
  func setFooterHeight(_ height: NSNumber) {
    let footerHeight = CGFloat(height.floatValue)
    guard let footerView, viewController.footerHeight != footerHeight else {
      return
    }

    viewController.footerHeight = footerHeight

    if footerView.subviews.first != nil {
      containerView?.bringSubviewToFront(footerView)
      footerConstraints?.height?.constant = viewController.footerHeight
    } else {
      containerView?.sendSubviewToBack(footerView)
      footerConstraints?.height?.constant = 0
    }

    if #available(iOS 15.0, *) {
      withPresentedSheet { _ in
        viewController.setupSizes()
      }
    }
  }

  @objc
  func setSizes(_ sizes: [Any]) {
    viewController.sizes = Array(sizes.prefix(3))

    if #available(iOS 15.0, *) {
      withPresentedSheet { _ in
        viewController.setupSizes()
      }
    }
  }

  @objc
  func setBackground(_ color: NSNumber?) {
    viewController.backgroundColor = RCTConvert.uiColor(color)
    viewController.setupBackground()
  }

  @objc
  func setBlurTint(_ tint: NSString?) {
    if let tint {
      viewController.blurEffect = UIBlurEffect(with: tint as String)
    } else {
      viewController.blurEffect = nil
    }

    viewController.setupBackground()
  }

  @objc
  func setCornerRadius(_ radius: NSNumber?) {
    var cornerRadius: CGFloat?

    if let radius {
      cornerRadius = CGFloat(radius.floatValue)
    }

    viewController.cornerRadius = cornerRadius
    if #available(iOS 15.0, *) {
      withPresentedSheet { sheet in
        sheet.preferredCornerRadius = viewController.cornerRadius
      }
    }
  }

  @objc
  func setGrabber(_ visible: Bool) {
    viewController.grabber = visible

    if #available(iOS 15.0, *) {
      withPresentedSheet { sheet in
        sheet.prefersGrabberVisible = visible
      }
    }
  }

  @objc
  func setDimmed(_ dimmed: Bool) {
    guard viewController.dimmed != dimmed else {
      return
    }

    viewController.dimmed = dimmed

    if #available(iOS 15.0, *) {
      withPresentedSheet { _ in
        viewController.setupDimmedBackground()
      }
    }
  }

  @objc
  func setDimmedIndex(_ index: NSNumber) {
    guard viewController.dimmedIndex != index.intValue else {
      return
    }

    viewController.dimmedIndex = index.intValue

    if #available(iOS 15.0, *) {
      withPresentedSheet { _ in
        viewController.setupDimmedBackground()
      }
    }
  }

  @objc
  func setScrollableHandle(_ tag: NSNumber?) {
    scrollView = uiManager?.view(forReactTag: tag)
  }

  // MARK: - Methods

  private func sizeInfoData(from sizeInfo: SizeInfo?) -> [String: Any]? {
    guard let sizeInfo else {
      return nil
    }

    return ["index": sizeInfo.index, "value": sizeInfo.value]
  }

  /// Use to customize some properties of the Sheet without fully reconfiguring.
  @available(iOS 15.0, *)
  func withPresentedSheet(completion: (UISheetPresentationController) -> Void) {
    guard isPresented, let sheet = viewController.sheetPresentationController else {
      return
    }

    sheet.animateChanges {
      completion(sheet)
    }
  }

  func dispatchEvent(name: String, block: RCTDirectEventBlock?, data: [String: Any]?) {
    // eventDispatcher doesn't work in New Arch so we need to call it directly :/
    // we needed eventDispatcher for Reanimated to work on old arch.
    #if RCT_NEW_ARCH_ENABLED
      block?(data)
    #else
      eventDispatcher?.send(TrueSheetEvent(viewTag: reactTag, name: name, data: data))
    #endif
  }

  func dismiss(promise: Promise) {
    guard isPresented else {
      promise.resolve(nil)
      return
    }

    viewController.dismiss(animated: true) {
      promise.resolve(nil)
    }
  }

  func present(at index: Int, promise: Promise?, animated: Bool = true) {
    let rvc = reactViewController()

    guard let rvc else {
      promise?.reject(message: "No react view controller present.")
      return
    }

    guard viewController.sizes.indices.contains(index) else {
      promise?.reject(message: "Size at \(index) is not configured.")
      return
    }

    if isPresented {
      withPresentedSheet { sheet in
        sheet.selectedDetentIdentifier = viewController.detentIdentifierForIndex(index)

        // Trigger onSizeChange event when size is changed while presenting
        viewControllerDidChangeSize(self.viewController.currentSizeInfo)
        promise?.resolve(nil)
      }
    } else {
      viewController.prepareForPresentation(at: index) {
        // Keep track of the active index
        self.activeIndex = index
        self.isPresented = true

        rvc.present(self.viewController, animated: animated) {
          if #available(iOS 15.0, *) {
            self.viewController.observeDrag()
          }

          let data = self.sizeInfoData(from: self.viewController.currentSizeInfo)
          self.dispatchEvent(name: "onPresent", block: self.onPresent, data: data)
          promise?.resolve(nil)
        }
      }
    }
  }
}
